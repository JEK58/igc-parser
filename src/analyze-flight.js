const turf = require('@turf/turf');

function analyzeFlight(flight, task) {
  let nextTP = 0;

  let start, finish;

  for (let i = 0; i < flight.length - 1; i++) {
    let fix1 = flight[i].coordinate;
    let fix2 = flight[i + 1].coordinate;

    if (nextTP < 2) {
      let point = task.points[0].observationZone.checkEnter(fix1, fix2);
      if (point) {
        nextTP = 1;
        start = { point, i, secOfDay: flight[i].secOfDay };
      }
    }

    if (nextTP === task.points.length - 1) {
      let point = task.points[nextTP].observationZone.checkEnter(fix1, fix2);
      if (point) {
        nextTP += 1;
        finish = { point, i, secOfDay: flight[i].secOfDay };
        break;
      }
    }

    let point = task.points[nextTP].observationZone.checkEnter(fix1, fix2);
    if (point) {
      nextTP += 1;
    }
  }

  return { start, finish };
}

module.exports = analyzeFlight;
