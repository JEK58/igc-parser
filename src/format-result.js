function formatResult(result) {
  return `Start: ${formatTime(result.start.secOfDay)}\nTime: ${formatTime(result.totalTime)}`;
}

function formatTime(secOfDay) {
  let sec = secOfDay % 60;
  let min = ((secOfDay - sec) / 60) % 60;
  let hour = (secOfDay - sec - min * 60) / 3600;
  return `${hour}:${min}:${sec}`;
}

module.exports = formatResult;
