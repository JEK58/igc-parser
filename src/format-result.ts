export function formatResult(result) {
  return `Start: ${formatTime(result.start.time)}\n` +
    `Time: ${formatTime(result.totalTime)}\n` +
    `Distance: ${(result.distance)} km\n` +
    `Speed: ${(result.speed)} km/h\n`;
}

export function formatTime(time) {
  let date = new Date(time);
  return date.toISOString();
}
