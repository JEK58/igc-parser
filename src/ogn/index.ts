import {Socket} from "net";
import * as readline from "readline";

export default class OGNClient {
  readonly host = 'aprs.glidernet.org';
  readonly port = 10152;

  readonly user = 'test';
  readonly pass = '-1';
  readonly appName = 'test';
  readonly appVersion = '1.0';

  readonly senders: string[];

  private readonly socket = new Socket();
  private readonly reader = readline.createInterface({ input: this.socket as NodeJS.ReadableStream });

  constructor(senders: string[]) {
    this.senders = senders;
    this.reader.on('line', line => this.handleLine(line));
  }

  connect() {
    this.socket.connect(this.port, this.host, () => {
      this.socket.write(`user ${this.user} pass ${this.pass} vers ${this.appName} ${this.appVersion}\n`);
      this.socket.emit('ready');
    });
  }

  handleLine(line: string) {
    let knownSender = this.senders.some(sender => line.indexOf(sender) !== -1);
    if (knownSender) {
      this.socket.emit('record', line);
    }
  }

  on(event: string, handler: Function) {
    if (event === 'line') {
      this.reader.on.call(this.reader, event, handler);
    } else {
      this.socket.on.call(this.socket, event, handler);
    }
  }
}
