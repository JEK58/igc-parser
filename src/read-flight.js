const fs = require('fs');

function readFlight(path) {
  return fs.readFileSync(path, 'utf8')
    .split('\n')
    .filter(line => line[0] === 'B')
    .map(line => line.match(/^B(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(\d{3})([NS])(\d{3})(\d{2})(\d{3})([EW])/))
    .filter(Boolean)
    .map(match => {
      let secOfDay = parseInt(match[1]) * 3600 + parseInt(match[2]) * 60 + parseInt(match[3]);
      let lat = parseInt(match[4]) + parseInt(match[5]) / 60 + parseInt(match[6]) / 60000;
      let lon = parseInt(match[8]) + parseInt(match[9]) / 60 + parseInt(match[10]) / 60000;
      let coordinate = [lon, lat];
      return { secOfDay, coordinate };
    });
}

module.exports = readFlight;
