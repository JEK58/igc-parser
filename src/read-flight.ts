import * as fs from "fs";

const RE_HFDTE = /^HFDTE(\d{2})(\d{2})(\d{2})/;
const RE_B = /^B(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(\d{3})([NS])(\d{3})(\d{2})(\d{3})([EW])/;

export interface Fix {
  secOfDay: number,
  time: number,
  coordinate: GeoJSON.Position,
}

export function readFlight(path: string): Fix[] {
  let lines = fs.readFileSync(path, 'utf8').split('\n');

  let date: number | undefined;
  let fixes: Fix[] = [];

  lines.forEach(line => {
    if (date === undefined) {
      date = findDate(line);
    }

    // TODO handle UTC-midnight wraparound
    if (date) {
      let fix = convertLine(line, date);
      if (fix) {
        fixes.push(fix);
      }
    }
  });

  return fixes;
}

function findDate(line: string): number | undefined {
  let match = line.match(RE_HFDTE);
  if (match) {
    let year = parseInt(match[3]) + 2000;
    let month = parseInt(match[2]) - 1;
    let day = parseInt(match[1]);
    return Date.UTC(year, month, day);
  }
}

function convertLine(line: string, date: number): Fix | undefined {
  let match = line.match(RE_B);
  if (!match) return;

  let secOfDay = parseInt(match[1]) * 3600 + parseInt(match[2]) * 60 + parseInt(match[3]);
  let time = date + secOfDay * 1000;

  let lat = parseInt(match[4]) + parseInt(match[5]) / 60 + parseInt(match[6]) / 60000;
  let lon = parseInt(match[8]) + parseInt(match[9]) / 60 + parseInt(match[10]) / 60000;
  let coordinate = [lon, lat];

  return { secOfDay, time, coordinate };
}
