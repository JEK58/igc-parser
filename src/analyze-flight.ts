import {Fix} from "./read-flight";
import Task from "./task/task";
import RacingTaskSolver from "./task/solver/racing-task-solver";
import AreaTaskSolver from "./task/solver/area-task-solver";

export function analyzeFlight(flight: Fix[], task: Task) {
  if (task.options.isAAT) {
    let analyzer = new AreaTaskSolver(task);
    analyzer.consume(flight);
    return analyzer.result;
  } else {
    let analyzer = new RacingTaskSolver(task);
    analyzer.consume(flight);
    return analyzer.result;
  }
}
