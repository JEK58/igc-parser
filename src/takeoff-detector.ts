import * as turf from '@turf/turf';

import {Fix} from "./read-flight";

export class TakeoffDetector {
  private _lastFix: Fix | undefined = undefined;

  onTakeoff: (fix: Fix) => void;
  onLanding: (fix: Fix) => void;

  flying: boolean = false;

  update(fix: Fix) {
    let speed = this._currentSpeed(fix);
    let flying = this._isFlightSpeed(speed);

    if (flying && !this.flying && this.onTakeoff) {
      this.onTakeoff(fix);
    }
    if (!flying && this.flying && this.onLanding) {
      this.onLanding(fix);
    }

    this.flying = flying;
    this._lastFix = fix;
  }

  _currentSpeed(fix): number {
    if (!this._lastFix) return 0;

    let distance = turf.distance(this._lastFix.coordinate, fix.coordinate);
    let milliseconds = fix.time - this._lastFix.time;
    let hours = milliseconds / 3600000;
    return distance / hours;
  }

  _isFlightSpeed(speed: number): boolean {
    return speed > 50;
  }
}
