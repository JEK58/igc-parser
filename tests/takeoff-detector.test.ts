import * as fs from "fs";

import {Fix, readFlight} from "../src/read-flight";
import {TakeoffDetector} from "../src/takeoff-detector";
import {formatTime} from "../src/format-result";

describe('TakeoffDetector', () => {
  fs.readdirSync(`${__dirname}/../fixtures/2017-07-15-lev`).filter(filename => (/\.igc$/i).test(filename)).forEach(filename => {
    let callsign = filename.match(/^(.{1,3})_/)![1];

    it(`detects takeoff and landing for "${callsign}"`, () => {
      let flight = readFlight(`${__dirname}/../fixtures/2017-07-15-lev/${filename}`);

      let detector = new TakeoffDetector();

      let takeoff: Fix, landing: Fix;
      detector.onTakeoff = fix => (takeoff = takeoff || fix);
      detector.onLanding = fix => (landing = fix);

      flight.forEach(fix => detector.update(fix));

      expect(`${takeoff && formatTime(takeoff.time)} - ${landing && formatTime(landing.time)}`).toMatchSnapshot();
    });
  });
});



