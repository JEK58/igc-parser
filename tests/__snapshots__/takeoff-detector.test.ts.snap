// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TakeoffDetector detects takeoff and landing for "1" 1`] = `"2017-07-15T08:38:00.000Z - 2017-07-15T14:52:48.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "1G" 1`] = `"2017-07-15T10:18:50.000Z - 2017-07-15T14:38:34.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "1R" 1`] = `"2017-07-15T10:11:24.000Z - 2017-07-15T14:58:03.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "33" 1`] = `"2017-07-15T09:43:50.000Z - 2017-07-15T14:37:32.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "999" 1`] = `"2017-07-15T10:23:03.000Z - 2017-07-15T15:11:31.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "E1" 1`] = `"2017-07-15T10:16:03.000Z - 2017-07-15T14:39:03.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "EI" 1`] = `"2017-07-15T10:12:41.000Z - 2017-07-15T14:51:29.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "FR" 1`] = `"2017-07-15T10:09:31.000Z - 2017-07-15T14:34:19.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "H1" 1`] = `"2017-07-15T10:07:09.000Z - 2017-07-15T12:04:25.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "IGP" 1`] = `"2017-07-15T10:02:19.000Z - 2017-07-15T14:39:31.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "JM" 1`] = `"2017-07-15T10:03:32.000Z - 2017-07-15T14:57:36.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "LC" 1`] = `"2017-07-15T09:58:02.000Z - 2017-07-15T15:14:26.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "MI" 1`] = `"2017-07-15T10:00:20.000Z - 2017-07-15T14:31:41.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "MS" 1`] = `"2017-07-15T09:52:44.000Z - 2017-07-15T15:00:08.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "P9" 1`] = `"2017-07-15T09:53:38.000Z - 2017-07-15T14:44:10.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "SW" 1`] = `"2017-07-15T09:50:10.000Z - 2017-07-15T14:33:47.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "T" 1`] = `"2017-07-15T09:42:34.000Z - 2017-07-15T14:35:41.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "T1" 1`] = `"2017-07-15T09:46:15.000Z - 2017-07-15T12:23:23.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "TA" 1`] = `"2017-07-15T09:40:37.000Z - 2017-07-15T14:27:15.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "WJ" 1`] = `"2017-07-15T09:38:28.000Z - 2017-07-15T14:36:51.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "XN" 1`] = `"2017-07-15T09:37:49.000Z - 2017-07-15T12:43:12.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "ZG" 1`] = `"2017-07-15T09:56:33.000Z - 2017-07-15T12:10:14.000Z"`;

exports[`TakeoffDetector detects takeoff and landing for "sq" 1`] = `"2017-07-15T09:48:09.000Z - 2017-07-15T12:32:05.000Z"`;
